distributed/shuffle/_worker_extension.py

```diff
@@ -362,9 +359,17 @@ def _() -> dict[str, list[tuple[ArrayRechunkShardID, bytes]]]:
             write the serialized payload directly to disk on the receiver.
             """
             out: dict[str, list[tuple[ArrayRechunkShardID, bytes]]] = defaultdict(list)
-            for id, nslice in self._slicing[input_partition]:
-                out[self.worker_for[id.chunk_index]].append(
-                    (id, pickle.dumps((id.shard_index, data[nslice])))
+            from itertools import product
+
+            shards = product(
```

Comment:
Since we only compute the splits _per axis_, we create the full cartesian product here again to determine each of the n-dimensional shards we have to slice from the input and send to the output.